Add support for ThinkPad T14*, T15*

ThinkPad T14*, T15* series is widely used and it would be great to have support on this platform as well. 

I was wondering how I could contribute in getting support on this platform?

After starting the built script on Ubuntu 24, this is the result of the command: 

without sudo:

```bash
$  ./target/release/lenovo-logo-changer 

[WARN ] Ignoring unknown button type: 
[INFO ] perform;
[INFO ] authenticate;
[INFO ] write_command; command=Auth(Some(External), Some([49, 48, 48, 48]))
[INFO ] write_commands; commands=[Auth(Some(External), Some([49, 48, 48, 48]))] extra_bytes=None
[INFO ] read_command;
[INFO ] read_commands; n_commands=1
[INFO ] send_secondary_commands;
[INFO ] write_commands; commands=[NegotiateUnixFD, Begin] extra_bytes=Some([108, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 109, 0, 0, 0, 1, 1, 111, 0, 21, 0, 0, 0, 47, 111, 114, 103, 47, 102, 114, 101, 101, 100, 101, 115, 107, 116, 111, 112, 47, 68, 66, 117, 115, 0, 0, 0, 2, 1, 115, 0, 20, 0, 0, 0, 111, 114, 103, 46, 102, 114, 101, 101, 100, 101, 115, 107, 116, 111, 112, 46, 68, 66, 117, 115, 0, 0, 0, 0, 3, 1, 115, 0, 5, 0, 0, 0, 72, 101, 108, 108, 111, 0, 0, 0, 6, 1, 115, 0, 20, 0, 0, 0, 111, 114, 103, 46, 102, 114, 101, 101, 100, 101, 115, 107, 116, 111, 112, 46, 68, 66, 117, 115, 0, 0, 0, 0])
[INFO ] receive_secondary_responses; expected_n_responses=1
[INFO ] read_commands; n_commands=1
[INFO ] new;
[INFO ] {}; task_name="org.a11y.Status proxy caching"
[INFO ] socket reader;
[INFO ] read_socket;
[INFO ] read_socket;
[INFO ] read_socket;
[INFO ] read_socket;
[INFO ] read_socket;
[INFO ] read_socket;
[INFO ] keep_updated;
[INFO ] Loading icon feature not supported on Linux
``` 

and with sudo: 

```bash
$  sudo ./target/release/lenovo-logo-changer 

[INFO ] Guessed window scale factor: 1
[ERROR] read lbldvc_var failed: variable not found: LBLDVC-871455d1-5576-4fb8-9865-af0824463c9f
[INFO ] Loading icon feature not supported on Linux
```

Rust version: 

```bash
$  rustc --version

rustc 1.75.0 (82e1608df 2023-12-21) (built from a source tarball)
```

Cargo version: 

```
$ cargo --version

cargo 1.92.0 (344c4567c 2025-10-21)
```

--- a/README.md
+++ b/README.md
@@ -86,6 +86,7 @@ All of the above operations need to be performed with administrator privileges.
 
 ## Supported Types
 
++ ThinkPad T14
 + ThinkBook 14 G4+ ARA
 + ThinkBook 16 G5+ ARP
 + IdeaPad Slim 5 14AHP9 (83DB)

--- a/src/lenlogo.rs
+++ b/src/lenlogo.rs
@@ -5,7 +5,7 @@ use std::path::Path;
 use std::str::FromStr;
 use efivar::efi::{Variable, VariableFlags};
 use sha2::{Sha256, Digest};
-use log::{debug, error, info};
+use log::{debug, error, info, warn};
 
 use crate::esp_partition::{copy_file_to_esp, delete_logo_path};
 
@@ -39,7 +39,12 @@ pub(crate) struct PlatformInfo {
     pub(crate) support: Vec<&'static str>,
     pub(crate) lbldesp_var: [u8; 10],
     pub(crate) lbldvc_var: [u8; 40],
+    pub(crate) lbldesp_var_name: String,
+    pub(crate) lbldvc_var_name: String,
 }
+
+const LBLDESP_VAR_NAME: &str = "LBLDESP-871455D0-5576-4FB8-9865-AF0824463B9E";
+const LBLDVC_VAR_NAME: &str = "LBLDVC-871455D1-5576-4FB8-9865-AF0824463C9F";
 
 impl Default for PlatformInfo {
     fn default() -> Self {
@@ -51,6 +56,8 @@ impl Default for PlatformInfo {
             support: Vec::new(),
             lbldesp_var: [0u8; 10],
             lbldvc_var: [0u8; 40],
+            lbldesp_var_name: LBLDESP_VAR_NAME.to_string(),
+            lbldvc_var_name: LBLDVC_VAR_NAME.to_string(),
         }
     }
 }
@@ -59,40 +66,106 @@ impl PlatformInfo {
     pub(crate) fn get_info(&mut self) -> bool {
         let varman = efivar::system();
 
-        let esp_var = Variable::from_str("LBLDESP-871455D0-5576-4FB8-9865-AF0824463B9E").unwrap();
+        let mut lbldesp_var_name = self.lbldesp_var_name.clone();
+        let mut lbldvc_var_name = self.lbldvc_var_name.clone();
+        let mut esp_var = Variable::from_str(&lbldesp_var_name).unwrap();
 
-        match varman.read(&esp_var) {
+        let esp_read_result = varman.read(&esp_var);
+        let (esp_buffer, _attr) = match esp_read_result {
             Ok((esp_buffer, _attr)) => {
-                if esp_buffer.len() != 10 {
-                    error!("read lbldesp_var failed: buffer length is not 10");
+                (esp_buffer, _attr)
+            }
+            Err(err) => {
+                #[cfg(target_os = "linux")]
+                {
+                    if let Some((desp_name, dvc_name)) = find_lenovo_vars() {
+                        lbldesp_var_name = desp_name;
+                        lbldvc_var_name = dvc_name;
+                        esp_var = Variable::from_str(&lbldesp_var_name).unwrap();
+                        match varman.read(&esp_var) {
+                            Ok((esp_buffer, _attr)) => (esp_buffer, _attr),
+                            Err(err) => {
+                                error!("read lbldesp_var failed: {}", err);
+                                return false;
+                            }
+                        }
+                    } else {
+                        error!("read lbldesp_var failed: {}", err);
+                        return false;
+                    }
+                }
+                #[cfg(not(target_os = "linux"))]
+                {
+                    error!("read lbldesp_var failed: {}", err);
                     return false;
                 }
-                self.enable = esp_buffer[0];
-                self.width = u32::from_le_bytes(esp_buffer[1..5].try_into().unwrap());
-                self.height = u32::from_le_bytes(esp_buffer[5..9].try_into().unwrap());
-                self.support = Self::support_format(esp_buffer[9]);
-                self.lbldesp_var = <[u8; 10]>::try_from(esp_buffer).unwrap();
-            },
-            Err(err) => {
-                error!("read lbldesp_var failed: {}", err);
-                return false;
+            }
+        };
+
+        if esp_buffer.len() != 10 {
+            error!("read lbldesp_var failed: buffer length is not 10");
+            return false;
+        }
+        self.enable = esp_buffer[0];
+        self.width = u32::from_le_bytes(esp_buffer[1..5].try_into().unwrap());
+        self.height = u32::from_le_bytes(esp_buffer[5..9].try_into().unwrap());
+        self.support = Self::support_format(esp_buffer[9]);
+        self.lbldesp_var = <[u8; 10]>::try_from(esp_buffer).unwrap();
+        self.lbldesp_var_name = lbldesp_var_name.clone();
+        self.lbldvc_var_name = lbldvc_var_name.clone();
+
+        #[cfg(target_os = "linux")]
+        {
+            if !efivar_exists(&lbldvc_var_name) {
+                if let Some(dvc_name) = find_lenovo_var("LBLDVC") {
+                    lbldvc_var_name = dvc_name;
+                } else {
+                    warn!("LBLDVC variable not found; proceeding without DVC support");
+                    self.lbldvc_var_name = String::new();
+                    return true;
+                }
             }
         }
 
-        let dvc_var = Variable::from_str("LBLDVC-871455D1-5576-4FB8-9865-AF0824463C9F").unwrap();
-        match varman.read(&dvc_var) {
-            Ok((dvc_buffer, _attr)) => {
-                if dvc_buffer.len() != 40 {
-                    error!("read lbldvc_var failed: buffer length is not 40");
+        let mut dvc_var = Variable::from_str(&lbldvc_var_name).unwrap();
+        let dvc_read_result = varman.read(&dvc_var);
+        let dvc_read_result = match dvc_read_result {
+            Ok((dvc_buffer, _attr)) => Some((dvc_buffer, _attr)),
+            Err(err) => {
+                #[cfg(target_os = "linux")]
+                {
+                    if let Some(dvc_name) = find_lenovo_var("LBLDVC") {
+                        lbldvc_var_name = dvc_name;
+                        dvc_var = Variable::from_str(&lbldvc_var_name).unwrap();
+                        match varman.read(&dvc_var) {
+                            Ok((dvc_buffer, _attr)) => Some((dvc_buffer, _attr)),
+                            Err(err) => {
+                                error!("read lbldvc_var failed: {}", err);
+                                return false;
+                            }
+                        }
+                    } else {
+                        warn!("LBLDVC variable not found; proceeding without DVC support");
+                        self.lbldvc_var_name = String::new();
+                        return true;
+                    }
+                }
+                #[cfg(not(target_os = "linux"))]
+                {
+                    error!("read lbldvc_var failed: {}", err);
                     return false;
                 }
-                self.version = u32::from_le_bytes(dvc_buffer[0..4].try_into().unwrap());
-                self.lbldvc_var = <[u8; 40]>::try_from(dvc_buffer).unwrap();
-            },
-            Err(err) => {
-                error!("read lbldvc_var failed: {}", err);
+            }
+        };
+
+        if let Some((dvc_buffer, _attr)) = dvc_read_result {
+            if dvc_buffer.len() != 40 {
+                error!("read lbldvc_var failed: buffer length is not 40");
                 return false;
             }
+            self.version = u32::from_le_bytes(dvc_buffer[0..4].try_into().unwrap());
+            self.lbldvc_var = <[u8; 40]>::try_from(dvc_buffer).unwrap();
+            self.lbldvc_var_name = lbldvc_var_name;
         }
         true
     }
@@ -117,10 +190,10 @@ impl PlatformInfo {
         // 修改logoinfo
         let mut esp_buffer = self.lbldesp_var.clone();
         esp_buffer[0] = 1;
-        let esp_var = Variable::from_str("LBLDESP-871455D0-5576-4FB8-9865-AF0824463B9E").unwrap();
+        let esp_var = Variable::from_str(&self.lbldesp_var_name).unwrap();
         
         // 在Linux下需要先移除immutable属性
-        let write_result = with_efi_var_writable("LBLDESP-871455d0-5576-4fb8-9865-af0824463b9e", || {
+        let write_result = with_efi_var_writable(&self.lbldesp_var_name, || {
             match varman.write(&esp_var, VariableFlags::from_bits(0x7).unwrap(), &esp_buffer) {
                 Ok(rt) => {
                     debug!("write lbldesp_var: {:?}", rt);
@@ -162,10 +235,14 @@ impl PlatformInfo {
     debug!("sha256_bytes: {:?}", sha256_bytes);
     debug!("dvc_buffer: {:?}", dvc_buffer);
 
-        let dvc_var = Variable::from_str("LBLDVC-871455D1-5576-4FB8-9865-AF0824463C9F").unwrap();
+        if self.lbldvc_var_name.is_empty() {
+            return true;
+        }
+
+        let dvc_var = Variable::from_str(&self.lbldvc_var_name).unwrap();
         
         // 在Linux下需要先移除immutable属性
-        let write_result = with_efi_var_writable("LBLDVC-871455d1-5576-4fb8-9865-af0824463c9f", || {
+        let write_result = with_efi_var_writable(&self.lbldvc_var_name, || {
             match varman.write(&dvc_var, VariableFlags::from_bits(0x7).unwrap(), &dvc_buffer) {
                 Ok(rt) => {
                     debug!("write lbldvc_var: {:?}", rt);
@@ -202,10 +279,10 @@ impl PlatformInfo {
         let mut esp_buffer = self.lbldesp_var.clone();
         if esp_buffer[0] != 0 {
             esp_buffer[0] = 0;
-            let esp_var = Variable::from_str("LBLDESP-871455D0-5576-4FB8-9865-AF0824463B9E").unwrap();
+            let esp_var = Variable::from_str(&self.lbldesp_var_name).unwrap();
             
             // 在Linux下需要先移除immutable属性
-            let write_result = with_efi_var_writable("LBLDESP-871455d0-5576-4fb8-9865-af0824463b9e", || {
+            let write_result = with_efi_var_writable(&self.lbldesp_var_name, || {
                 match varman.write(&esp_var, VariableFlags::from_bits(0x7).unwrap(), &esp_buffer) {
                     Ok(rt) => {
                         debug!("write lbldesp_var: {:?}", rt);
@@ -229,12 +306,16 @@ impl PlatformInfo {
         }
 
         // 修改logocheck
+        if self.lbldvc_var_name.is_empty() {
+            return status;
+        }
+
         let mut dvc_buffer = self.lbldvc_var.clone();
         if dvc_buffer[4..40] != [0u8; 36] {
             dvc_buffer[4..40].clone_from_slice(&[0u8; 36]);
-            let dvc_var = Variable::from_str("LBLDVC-871455D1-5576-4FB8-9865-AF0824463C9F").unwrap();
+            let dvc_var = Variable::from_str(&self.lbldvc_var_name).unwrap();
             
-            let write_result = with_efi_var_writable("LBLDVC-871455d1-5576-4fb8-9865-af0824463c9f", || {
+            let write_result = with_efi_var_writable(&self.lbldvc_var_name, || {
                 match varman.write(&dvc_var, VariableFlags::from_bits(0x7).unwrap(), &dvc_buffer) {
                     Ok(rt) => { debug!("write lbldvc_var: {:?}", rt); Ok(()) },
                     Err(err) => Err(format!("write lbldvc_var failed: {}", err))
@@ -277,6 +358,59 @@ impl PlatformInfo {
         }
         support_types
     }
+}
+
+#[cfg(target_os = "linux")]
+fn find_lenovo_vars() -> Option<(String, String)> {
+    let entries = std::fs::read_dir("/sys/firmware/efi/efivars").ok()?;
+    let mut lbldesp_var_name = None;
+    let mut lbldvc_var_name = None;
+
+    for entry in entries.flatten() {
+        let file_name = entry.file_name();
+        let file_name = file_name.to_string_lossy();
+        let upper = file_name.to_ascii_uppercase();
+
+        if lbldesp_var_name.is_none() && upper.starts_with("LBLDESP") {
+            lbldesp_var_name = Some(file_name.to_string());
+        } else if lbldvc_var_name.is_none() && upper.starts_with("LBLDVC") {
+            lbldvc_var_name = Some(file_name.to_string());
+        }
+
+        if lbldesp_var_name.is_some() && lbldvc_var_name.is_some() {
+            break;
+        }
+    }
+
+    match (lbldesp_var_name, lbldvc_var_name) {
+        (Some(desp), Some(dvc)) => Some((desp, dvc)),
+        _ => None,
+    }
+}
+
+#[cfg(target_os = "linux")]
+fn find_lenovo_var(prefix: &str) -> Option<String> {
+    let entries = std::fs::read_dir("/sys/firmware/efi/efivars").ok()?;
+
+    for entry in entries.flatten() {
+        let file_name = entry.file_name();
+        let file_name = file_name.to_string_lossy();
+        let upper = file_name.to_ascii_uppercase();
+        if upper.starts_with(&prefix.to_ascii_uppercase()) {
+            return Some(file_name.to_string());
+        }
+    }
+
+    None
+}
+
+#[cfg(target_os = "linux")]
+fn efivar_exists(var_name: &str) -> bool {
+    if var_name.is_empty() {
+        return false;
+    }
+    let path = Path::new("/sys/firmware/efi/efivars").join(var_name);
+    path.exists()
 }
 
 fn calculate_sha256(file_path: &str) -> io::Result<String> {

--- a/src/main.rs
+++ b/src/main.rs
@@ -46,6 +46,8 @@ struct LogoOperationResult {
     enable: u8,
     lbldesp_var: [u8; 10],
     lbldvc_var: [u8; 40],
+    lbldesp_var_name: String,
+    lbldvc_var_name: String,
 }
 
 // 在后台线程执行设置Logo操作
@@ -54,6 +56,8 @@ fn perform_set_logo_operation(
     show_loading_icon: bool,
     lbldesp_var: [u8; 10],
     lbldvc_var: [u8; 40],
+    lbldesp_var_name: String,
+    lbldvc_var_name: String,
 ) -> LogoOperationResult {
     // 先设置加载图标
     let loading_icon_result = NativePlatform::set_loading_icon(show_loading_icon);
@@ -65,6 +69,8 @@ fn perform_set_logo_operation(
     let mut temp_info = PlatformInfo::default();
     temp_info.lbldesp_var = lbldesp_var;
     temp_info.lbldvc_var = lbldvc_var;
+    temp_info.lbldesp_var_name = lbldesp_var_name;
+    temp_info.lbldvc_var_name = lbldvc_var_name;
     temp_info.width = u32::from_le_bytes(lbldesp_var[1..5].try_into().unwrap());
     temp_info.height = u32::from_le_bytes(lbldesp_var[5..9].try_into().unwrap());
     
@@ -82,6 +88,8 @@ fn perform_set_logo_operation(
         enable: updated_info.enable,
         lbldesp_var: updated_info.lbldesp_var,
         lbldvc_var: updated_info.lbldvc_var,
+        lbldesp_var_name: updated_info.lbldesp_var_name,
+        lbldvc_var_name: updated_info.lbldvc_var_name,
     }
 }
 
@@ -89,6 +97,8 @@ fn perform_set_logo_operation(
 fn perform_restore_logo_operation(
     lbldesp_var: [u8; 10],
     lbldvc_var: [u8; 40],
+    lbldesp_var_name: String,
+    lbldvc_var_name: String,
 ) -> LogoOperationResult {
     // 设置加载图标为启用
     let loading_icon_result = NativePlatform::set_loading_icon(true);
@@ -100,6 +110,8 @@ fn perform_restore_logo_operation(
     let mut temp_info = PlatformInfo::default();
     temp_info.lbldesp_var = lbldesp_var;
     temp_info.lbldvc_var = lbldvc_var;
+    temp_info.lbldesp_var_name = lbldesp_var_name;
+    temp_info.lbldvc_var_name = lbldvc_var_name;
     
     let success = temp_info.restore_logo();
     
@@ -115,6 +127,8 @@ fn perform_restore_logo_operation(
         enable: updated_info.enable,
         lbldesp_var: updated_info.lbldesp_var,
         lbldvc_var: updated_info.lbldvc_var,
+        lbldesp_var_name: updated_info.lbldesp_var_name,
+        lbldvc_var_name: updated_info.lbldvc_var_name,
     }
 }
 
@@ -216,7 +230,7 @@ impl MyApp {
                 }
 
                 if let Some(picked_path) = &self.picked_path  {
-                    if self.platform_info.version == 0x20003 {
+                    if self.platform_info.version == 0x20003 || self.platform_info.lbldvc_var_name.is_empty() {
                         ui.horizontal(|ui| {
                             ui.label(t(lang, "picked_image").as_ref());
                             ui.monospace(picked_path);
@@ -236,10 +250,19 @@ impl MyApp {
                             let show_loading_icon = self.set_loading_icon;
                             let lbldesp_var = self.platform_info.lbldesp_var;
                             let lbldvc_var = self.platform_info.lbldvc_var;
+                            let lbldesp_var_name = self.platform_info.lbldesp_var_name.clone();
+                            let lbldvc_var_name = self.platform_info.lbldvc_var_name.clone();
                             
                             // 在后台线程执行操作
                             self.set_logo_promise = Some(Promise::spawn_thread("set_logo", move || {
-                                perform_set_logo_operation(img_path, show_loading_icon, lbldesp_var, lbldvc_var)
+                                perform_set_logo_operation(
+                                    img_path,
+                                    show_loading_icon,
+                                    lbldesp_var,
+                                    lbldvc_var,
+                                    lbldesp_var_name,
+                                    lbldvc_var_name,
+                                )
                             }));
                             self.pending_set_logo = false;
                         }
@@ -256,6 +279,8 @@ impl MyApp {
                                 self.platform_info.enable = result.enable;
                                 self.platform_info.lbldesp_var = result.lbldesp_var;
                                 self.platform_info.lbldvc_var = result.lbldvc_var;
+                                self.platform_info.lbldesp_var_name = result.lbldesp_var_name.clone();
+                                self.platform_info.lbldvc_var_name = result.lbldvc_var_name.clone();
                                 
                                 // 清除Promise
                                 self.set_logo_promise = None;
@@ -292,10 +317,17 @@ impl MyApp {
                     // 捕获需要的数据
                     let lbldesp_var = self.platform_info.lbldesp_var;
                     let lbldvc_var = self.platform_info.lbldvc_var;
+                    let lbldesp_var_name = self.platform_info.lbldesp_var_name.clone();
+                    let lbldvc_var_name = self.platform_info.lbldvc_var_name.clone();
                     
                     // 在后台线程执行操作
                     self.restore_logo_promise = Some(Promise::spawn_thread("restore_logo", move || {
-                        perform_restore_logo_operation(lbldesp_var, lbldvc_var)
+                        perform_restore_logo_operation(
+                            lbldesp_var,
+                            lbldvc_var,
+                            lbldesp_var_name,
+                            lbldvc_var_name,
+                        )
                     }));
                     self.pending_restore_logo = false;
                 }
@@ -312,6 +344,8 @@ impl MyApp {
                         self.platform_info.enable = result.enable;
                         self.platform_info.lbldesp_var = result.lbldesp_var;
                         self.platform_info.lbldvc_var = result.lbldvc_var;
+                        self.platform_info.lbldesp_var_name = result.lbldesp_var_name.clone();
+                        self.platform_info.lbldvc_var_name = result.lbldvc_var_name.clone();
                         self.is_support = result.enable != 0 || result.success;
                         
                         // 清除Promise

